src/Nonlinear/Nonlinear.jl

```diff
@@ -0,0 +1,509 @@
+module Nonlinear
+
+import Base.Meta: isexpr
+import ForwardDiff
+import ..MathOptInterface
+import OrderedCollections: OrderedDict
+import SparseArrays
+
+using SpecialFunctions
+
+const MOI = MathOptInterface
+
+# Override basic math functions to return NaN instead of throwing errors.
+# This is what NLP solvers expect, and sometimes the results aren't needed
+# anyway, because the code may compute derivatives wrt constants.
+import NaNMath:
+    sin,
+    cos,
+    tan,
+    asin,
+    acos,
+    acosh,
+    atanh,
+    log,
+    log2,
+    log10,
+    lgamma,
+    log1p,
+    pow,
+    sqrt
+
+include("univariate_expressions.jl")
+include("operators.jl")
+include("types.jl")
+include("parse.jl")
+
+include("ReverseAD/ReverseAD.jl")
+
+"""
+    set_objective(data::NonlinearData, obj)::Nothing
```

Comment:
```suggestion
    set_objective(data::NonlinearData, obj::Expr)::Nothing
```
could help users, with a sentence to say what obj must be

Reply:
I've updated the docstrings. This can be something other than `Expr`.